test_runner/batch_others/test_proxy.py

```diff
@@ -22,6 +26,99 @@ def test_password_hack(static_proxy):
+async def test_psql_session_id(vanilla_pg, link_proxy):
+    """
+    Test copied and modified from: test_project_psql_link_auth test from cloud/tests_e2e/tests/test_project.py
+    This is only one half of the test: the half that establishes connection to the proxy and retrieves the session_id
+    The second half of the test required more heavy machinery to be ported from the cloud codebase
+        (such as: project_state(..), and psql_project_connect)
+    to connect to postgres and make sure the end-to-end behavior is satisfied.
+    """
+    psql = PSQL(
+        host=link_proxy.host,
+        port=link_proxy.proxy_port,
+    )
+    proc = await psql.run()
+
+    # Read line by line output of psql.
+    # We expect to see result in first 3 lines. [this is flexible, change if need be]
+    # Expected output:
+    #   A notice that contains a welcome to neon message that contains a http url in the form of
+    #   "http://dummy-uri/<psql_session_id>"
+    #   (here dummy-uri was passed from the neon_fixture that created the link_proxy
+    #   The url has to be on its own line.
+    #   Example:
+    #   "
+    #   NOTICE:  Welcome to Neon!
+    #   Authenticate by visiting:
+    #      http://dummy-uri/<psql_session_id>
+    #   "
+
+    max_num_lines_of_welcome_message = 15
+    attempt = 0
+    psql_session_id = None
+
+    log.info(f"Neon Welcome Message:")
+    for attempt in range(max_num_lines_of_welcome_message):
+        raw_line = await proc.stderr.readline()
+        if not raw_line:
+            # output ended
+            break
+        log.info(f"{raw_line}")
+        line = raw_line.decode("utf-8").strip()
+        if line.startswith("http"):
+            line_parts = line.split("/")
+            psql_session_id = line_parts[-1]
+            link_auth_uri = '/'.join(line_parts[:-1])
+            assert link_auth_uri == link_proxy.link_auth_uri, \
+                f"Line='{line}' should contain a http auth link of form '{link_proxy.link_auth_uri}/<psql_session_id>'."
+            break
+        log.debug("line %d does not contain expected result: %s", attempt, line)
+
+    assert attempt <= max_num_lines_of_welcome_message, "exhausted given attempts, did not get the result"
+    assert psql_session_id is not None, "psql_session_id not found from output of proc.stderr.readline()"
+    log.debug(f"proc.stderr.readline() #{attempt} has the result: {psql_session_id=}")
+    log.info(f"proc.stderr.readline() #{attempt} has the result: {psql_session_id=}")
+
+    # make a db and send to proxy.
+    # greate a postgres instance and give the proxy the info to send queries to it.
+
+    pg_user = "proxy"
+    pg_password = "password"
+
+    vanilla_pg.start()
+    vanilla_pg.safe_psql("create user "+pg_user+" with login superuser password '"+pg_password+"'")
+
+    port = vanilla_pg.default_options['port']
+    host = vanilla_pg.default_options['host']
+    dbname = vanilla_pg.default_options['dbname']
+
+    cmd_line_args__to__mgmt = [
```

Comment:
No need in calling `psql` directly. There is a `psycopg2`

Reply:
could you please guide me more in this, how do I use psycpg2 to do the same thing?

Reply:
Instantiate a `PgProtocol` on this port and send the `db_info` via `.safe_psql` (poorly named function, it's not psql, it uses psycopg2). Or see how `PgProtocol` uses `psycopg2` and to the same directly, if that's easier. We're aiming for maximal readability of the code at the end

The bottom line is: Earlier you absolutely needed **psql** to connect to the **main port** of the proxy because you need a client that will print out the notice and allow you to parse the session id. But here you're connecting to the **mgmt endpoint** of the proxy and you don't have such a limitation, you can use any postgres client you want. Using PgProtocol or psycopg2 is simpler than creating a psql process.

Slack me if you get stuck

Reply:
> Earlier you absolutely needed psql to connect to the main port of the proxy because you need a client that will print out the notice and allow you to parse the session id.

BTW, can't we use https://www.psycopg.org/docs/connection.html#connection.notices ?

Reply:
You need to connect to create a connection object to get the notices. We need the notice before the client gets ReadyForQuery

Reply:
Ah, gotcha

Reply:
So I'm not exactly sure how to use psycopg2 here, this is what I have so far (it's not working, failing with error 
> E       psycopg2.DatabaseError: expected value at line 1 column 1

when it tries to connect.

What I have so far:

```
    db_info_dict = {
        "session_id": psql_session_id,
        "result": {
            "Success": {
                "host": host,
                "port": port,
                "dbname": dbname,
                "user": pg_user,
                "password": pg_password
            }
        }
    }

    mgmt_port_str = str(7000)  # hardcoded from proxy's backend I think
    db_info_str = json.dumps(db_info_dict)

    with closing(psycopg2.connect(port = mgmt_port_str, host = "127.0.0.1")) as conn:
      with conn.cursor() as cur:
          query = db_info_str
          log.info(f"Executing query: {query}")
          cur.execute(query)
          assert cur.description is not None
          assert "ok" in cur.fetchall()
```
but I'm not sure what params to call psycopg2 with. The ones here  (port = mgmt_port_str, host = "127.0.0.1") don't seem correct because the mgmt is not a databse. Any thoughts on this? Any suggestions how to set up the call to psycopg2?

Reply:
Discussed offline. Apparently psycopg2 returns some internal error (`expected value at line 1 column 1`), that's not easy to debug since we don't get a stack trace (it's a c library?)

IMO it's not a big deal since the mgmt endpoint is deprecated and we'll delete this soon. Or let's continue the conversation here https://github.com/neondatabase/neon/pull/2316 and move on in this PR.